server/src/core/search/search_executor/preprocess.rs

```diff
@@ -0,0 +1,297 @@
+            // end of quotes
+            let raw_token = q.get(token_start..i + c.len_utf8());
+            if let Some(token) = raw_token {
+                tokens.push(InputToken {
+                    // Note: The trim_end also trims within unclosed quotes at the end of the query,
+                    //       but currently I don't think this is an issue.
+                    s: token.trim_end().to_lowercase(),
+                    regular_split: true,
+                    // `closed` indicates whether the token has been closed (by whitespace or quote)
+                    // at the end, when this is the last token. This is relevant because MeiliSearch
+                    // treats whitespace at the end differently, and we might want to imitate that
+                    // behaviour.
+                    closed: !(i + c.len_utf8() == q.len()
+                        && (within_quotes || (c != '"' && !c.is_whitespace()))),
+                });
+            }
+
+            token_start = i + 1;
+        } else if !within_quotes && c.is_whitespace() {
+            //    ^
+            // To avoid empty tokens when there are multiple whitespaces, we need
+            // to move the token start even if there was no split.
+            token_start = i + 1;
+        }
+
+        if c.is_alphabetic() {
+            alphabetic_counter += 1;
+        } else {
+            alphabetic_counter = 0;
+        }
+    }
+
+    tokens
+}
+
+#[cfg(test)]
+mod tokenizer_tests {
+    use super::*;
+
+    fn reg(s: &str) -> InputToken {
+        InputToken {
+            s: s.to_string(),
+            regular_split: true,
+            closed: true,
+        }
+    }
+
+    fn irreg(s: &str) -> InputToken {
+        InputToken {
+            s: s.to_string(),
+            regular_split: false,
+            closed: true,
+        }
+    }
+
+    fn assert_token(q: String, expected: Vec<InputToken>) {
+        assert_eq!(
+            tokenize_input_query(q.clone()),
+            expected,
+            "tokenization for '{}' failed",
+            q
+        );
+    }
+
+    fn assert_tokens(q: &str, mut expected: Vec<InputToken>) {
+        let sqs = format!(" {} ", q);
+        let qs = format!("{} ", q);
+        assert_token(sqs, expected.clone());
+        assert_token(qs, expected.clone());
+        if !expected.is_empty() {
+            let mut last = expected.pop().unwrap();
+            last.closed = false;
+            expected.push(last);
+        }
+        let sq = format!(" {}", q);
+        assert_token(sq, expected.clone());
+        assert_token(q.to_string(), expected.clone());
+    }
+
+    #[macro_export]
+    macro_rules! assert_identical {
+        ($x:expr) => {
+            assert_tokens($x, vec![reg($x)]);
+        };
+    }
+
+    #[test]
+    fn empty() {
+        assert_tokens("", vec![]);
+        assert_tokens("\t", vec![]);
+        assert_tokens("\n", vec![]);
+    }
+
+    #[test]
+    fn quoting() {
+        assert_identical!("\"");
```

Comment:
Not sure what the correct behaviour should be here.

The reason is that on opening quote only sets `within_quotes` to true, but does not mark a 'split', that means it doesn't create a new token. Only a whitespace or the end of the string marks a (regular) split. That means just an opening quote at the end is ignored.

The relevant code is:
```rust
    for (i, c) in q.char_indices() {
        // Quote escaping is not supported
        if c == '"' {
            within_quotes = !within_quotes;

            // Only closing (even-numbered) quotes do a split (see below).
            // Opening quotes should not split (therefore the continue).
            if within_quotes {
                continue;
            }
        }

        // [... some code here in between, another if ...]
        } else if !within_quotes && c.is_whitespace() {
            //    ^
            // To avoid empty tokens when there are multiple whitespaces, we need
            // to move the token start even if there was no split.
            token_start = i + 1;
        }

```

Visualizing what happens (`*` is `token_start`, `^` is the current character position):
```
Variant (a)
'abc "'
 ^
 *
'abc "'
 *^
'abc "'
 * ^
'abc "'  // 'abc' is split out as new token, token_start (for the next token) is set to the next char after the whitespace
    ^*
'abc "'  // `within_quotes` is now set to true, end of string is reached, but no end of string split is performed, because of the `continue`
     ^
     *

Variant (b)
'abc " '   // `within_quotes` is now set to true, but nothing more than that happens
     ^
     *
'abc " '   // end of string is reached and an end of string split is performed (not because of the whitespace because we're in quotes) making '"' a new token 
      ^
       *
```

Again, it probably depends on what behavior we want. I think it makes more sense to ignore the opening quote at the end of a string, because I don't know why we should pass it to MeiliSearch (it doesn't have any effect).

Reply:
i.e. my ideal would be 

```
#[test]
    fn quoting() {
        assert_tokens("\"", vec![]);
        assert_tokens("\"\"", vec![]);
        assert_tokens("\"\"", vec![]);
        assert_tokens("\" \"\"", vec![reg("\" \"")]);
        assert_identical!("\"a\"");
        assert_identical!("\"a \"");
        assert_identical!("\"a a \"");
        assert_identical!("\" a a \"");
    }
```

Reply:
Though all in all quoted search is a special case that people do deliberately, so they expect correct results only when they finished typing.

Reply:
How about this compromise:
I migrate the quote-excapes code into another testcase, mark that castcase with `#[ignore]` and `//TODO: test what happens when you type with quotes autoclosed (just whether it's irritating while typing)` and an Issue, to get this PR finally done?

Reply:
The deeper reasoning is, that #115 #114 #106 #103 #81 and #39 (the real reason) are depending (at least I think of them that way) on this PR..

Reply:
I just tested it, and it appears MeiliSearch does autoclose quotes as well, so autoclosing in our tokenizer shoudn't have side-effects with MeiliSearch. So we can use your test cases above. I'm sorry for the delay.

Reply:
It's implemented in https://github.com/TUM-Dev/navigatum/pull/104/commits/848eda3c86711be8c9b387db2fa219b6142c05e6 now

Reply:
Thanks :)